File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/bytecode/CoreMixinCV.java
package net.yakclient.mixin.base.internal.bytecode;

import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;

import java.util.Map;
import java.util.Queue;

public class CoreMixinCV extends MixinCV {
    public CoreMixinCV(ClassVisitor visitor, Map<String, Queue<QualifiedInstruction>> injectors) {
        super(visitor, injectors);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        final MethodVisitor visitor = super.visitMethod(access, name, desc, signature, exceptions);

        final var qualifiedName = name + desc;
        if (this.hasInjection(qualifiedName)) {
            var last = visitor;

            final Queue<QualifiedInstruction> instructions = this.getInjection(qualifiedName);
            for (QualifiedInstruction insn : instructions) {
                last = MixinPatternMatcher.corePatternMatcher(insn.getInjectionType(), last, insn.getInsn());
            }

            return last;
        }

        return visitor;
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/bytecode/CoreMixinCV.kt
package net.yakclient.mixins.base.internal.bytecode

import net.yakclient.mixins.base.internal.methodadapter.MixinPatternMatcher
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import java.util.*

class CoreMixinCV(
    visitor: ClassVisitor?,
    injectors: Map<String, Queue<QualifiedInstruction>>
) : MixinCV(visitor, injectors) {
    override fun visitMethod(
        access: Int,
        name: String,
        desc: String,
        signature: String?,
        exceptions: Array<String>?
    ): MethodVisitor {
        val visitor = super.visitMethod(access, name, desc, signature, exceptions)
        val qualifiedName = name + desc
        if (hasInjection(qualifiedName)) {
            var last = visitor
            val instructions = getInjection(qualifiedName)
            for (insn in instructions) {
                last = MixinPatternMatcher.createCoreNode(insn.injectionType, last, insn.insn)
            }
            return last
        }
        return visitor
    }
}